gpytorch/functions/_inv_quad_logdet.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
+
+import warnings
+
+import torch
+from torch.autograd import Function
+
+from .. import settings
+from ..utils.lanczos import lanczos_tridiag_to_diag
+from ..utils.stochastic_lq import StochasticLQ
+
+
+class InvQuadLogdet(Function):
+    """
+    Given a PSD matrix A (or a batch of PSD matrices A), this function computes one or both
+    of the following
+    - The matrix solves A^{-1} b
+    - logdet(A)
+
+    This function uses preconditioned CG and Lanczos quadrature to compute the inverse quadratic
+    and log determinant terms, using the variance reduction strategy outlined in:
+    ``Reducing the Variance of Gaussian Process Hyperparameter Optimization with Preconditioning''
+    (https://arxiv.org/abs/2107.00243)
+    """
+
+    @staticmethod
+    def forward(
+        ctx,
+        representation_tree,
+        precond_representation_tree,
+        preconditioner,
+        num_precond_args,
+        inv_quad,
+        probe_vectors,
+        probe_vector_norms,
+        *args,
+    ):
+        """
+        *args - The arguments representing the PSD matrix A (or batch of PSD matrices A)
+        If self.inv_quad is true, the first entry in *args is inv_quad_rhs (Tensor)
+        - the RHS of the matrix solves.
+
+        Returns:
+        - (Scalar) The inverse quadratic form (or None, if self.inv_quad is False)
+        - (Scalar) The log determinant (or None, self.if logdet is False)
+        """
+
+        ctx.representation_tree = representation_tree
+        ctx.precond_representation_tree = precond_representation_tree
+        ctx.preconditioner = preconditioner
+        ctx.inv_quad = inv_quad
+        ctx.num_precond_args = num_precond_args
+
+        matrix_args = None
+        precond_args = tuple()
+        inv_quad_rhs = None
+        if ctx.inv_quad:
+            inv_quad_rhs = args[0]
+            args = args[1:]
+        if ctx.num_precond_args:
+            matrix_args = args[:-num_precond_args]
+            precond_args = args[-num_precond_args:]
+        else:
+            matrix_args = args
+
+        # Get closure for matmul
+        lazy_tsr = ctx.representation_tree(*matrix_args)
+        precond_lt = ctx.precond_representation_tree(*precond_args)
+
+        # Get info about matrix
+        ctx.dtype = lazy_tsr.dtype
+        ctx.device = lazy_tsr.device
+        ctx.matrix_shape = lazy_tsr.matrix_shape
+        ctx.batch_shape = lazy_tsr.batch_shape
+
+        # Probe vectors
+        if probe_vectors is None or probe_vector_norms is None:
+            num_random_probes = settings.num_trace_samples.value()
+            if settings.deterministic_probes.on():
```

Comment:
Should we deprecate deterministic_probes? Variance reduction more properly solved the problem this was originally trying to, and as far as I can tell this isn't used anywhere.

Reply:
Done. The deprecation warning links to this PR.